### ISSUE TITLE
CACD preprocess script doesn't write csv with only centered entries

### ISSUE THREAD

**Issue by qwitwa** [Raschka-research-group/coral-cnn#2]
The dlib face detector can fail to find only 1 face in an image, in which case no image is saved in the centred image folder. The model code for CACD seems to reference some sort of output csv from this process, but the keep_picture list in preprocess_cacd.py is never actually written anywhere, never mind into a CSV file. Using the existing csv files from this repo fails because some of the images referenced no longer exist after the centering process.

**Comment by rasbt** [comment]
Did you use the `/datasets/image-preprocessing-code/preprocess-cacd.py` for the CACD preprocessing? I am pretty sure that I created the dataset splits based on the images that were left over from this preprocessing step. 

> The model code for CACD seems to reference some sort of output csv from this process, but the keep_picture list in preprocess_cacd.py is never actually written anywhere, never mind into a CSV file.

you can find the CSV files here:
- datasets/train-test-csv/cacd_train.csv
- datasets/train-test-csv/cacd_test.csv

they were generated by running the script mentioned above then created those based on the images written to the output folder by randomly partitioning them into training and test sets.

Please let me know if there are any issues with the .csv files above.

**Comment by TZZZZ** [comment]
I have the same problem. It seems what in some files in ./datasets/train-test-csv/cacd_test.csv aligner for CACD was not successful and it skips them. Maybe libraries we updated from that time. Should I change some parameters in aligner? 

I have problems with files

38_Taraji_P._Henson_0005.jpg
54_Ang_Lee_0016.jpg
34_Emily_Mortimer_0005.jpg
30_Cam_Gigandet_0003.jpg
27_Khloé_Kardashian_0008.jpg
51_Angela_Bassett_0011.jpg
58_Kate_Capshaw_0004.jpg
25_Rebecca_Hall_0002.jpg
46_Diane_Lane_0008.jpg
45_Luke_Goss_0005.jpg
37_Eric_Dane_0006.jpg
33_Zachary_Levi_0006.jpg
48_Kevin_Pollak_0010.jpg
46_Jeanne_Tripplehorn_0008.jpg
49_Teri_Hatcher_0011.jpg
42_Jason_Isaacs_0007.jpg
34_Cameron_Diaz_0015.jpg
57_Tony_Shalhoub_0006.jpg
32_Jennifer_Carpenter_0007.jpg
52_Brendan_Gleeson_0010.jpg
43_Pamela_Adlon_0009.jpg
25_Sarah_Shahi_0010.jpg
53_Kristine_DeBell_0012.jpg
57_Charles_Martin_Smith_0017.jpg
54_Marg_Helgenberger_0012.jpg
43_Ralph_Fiennes_0017.jpg
37_Victor_Webster_0008.jpg
38_Corey_Haim_0008.jpg
39_Carmen_Serano_0003.jpg
53_Isabelle_Adjani_0002.jpg
39_Ethan_Hawke_0015.jpg
35_Eminem_0013.jpg
15_Juno_Temple_0001.jpg
26_Michael_Angarano_0001.jpg
49_Michael_Jackson_0018.jpg
45_Bai_Ling_0002.jpg
53_Robert_Wisdom_0003.jpg
37_Hugh_Jackman_0007.jpg
45_Andre_Braugher_0009.jpg
25_Christy_Romano_0010.jpg
57_Michael_Emerson_0001.jpg
20_Dianna_Agron_0013.jpg
34_Russell_Brand_0008.jpg
17_Margot_Robbie_0013.jpg
53_Sarah_Douglas_0017.jpg
32_Joseph_Gordon-Levitt_0004.jpg
26_Brooklyn_Decker_0016.jpg
36_Anne_Heche_0009.jpg
28_Katie_Holmes_0002.jpg
45_Lauren_Holly_0003.jpg
53_Ned_Luke_0013.jpg

**Comment by rasbt** [comment]
Hm, weird. Let me reopen this and I will try to get to this later (busy day today, have to prepare lectures and a talk, but will try to look into this/send you the files ~tomorrow)

**Comment by rasbt** [comment]
Regarding the issue why the images couldn't be cropped ... that's a tricky one. I suspect `dlib.get_frontal_face_detector()` is the culprit, but why it behaves differently could be due to

- different dlib version
- different Python version (particularly 2.7 vs 3.7)
- different compilers used for building dlib
- ...

In any case, I shared all the missing images you listed here:

https://drive.google.com/drive/folders/1OLsUdaSEQgN9YzTICYTPkh0kOxJQWqZQ?usp=sharing

**Comment by TZZZZ** [comment]
Thank you! Now I have all aligned files from CACD test. I tested the network by applying cacd single-image-prediction for each file from ./datasets/train-test-csv/cacd_test.csv and I got result MAE = 5.8199 instead of MAE = 5.56 from the article. Do you know what could be the problem?

**Comment by rasbt** [comment]
Glad that it works overall! Regarding the MAE, which model are you running? Also, depending on the random seed, there are differences from run to run. Lastly, it was run on PyTorch 1.1. It could be that running the code in PyTorch 1.2 or 1.3 results in small differences as well. It would be interesting to know if all models have slightly lower MAE or whether this is the only one.
